java-squid/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValueRelation.java

```diff
@@ -17,93 +17,126 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
-package org.sonar.java.se;
+package org.sonar.java.se.symbolicvalues;
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.java.se.ConstraintManager.BooleanConstraint;
+import org.sonar.java.se.ProgramState;
 
 import javax.annotation.CheckForNull;
 
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
-import java.util.HashSet;
+import java.util.Comparator;
 import java.util.List;
 import java.util.Set;
+import java.util.TreeSet;
+
+public abstract class SymbolicValueRelation extends BinarySymbolicValue {
+
+  static final String TRANSITIVE_RELATIONS_EXCEEDED = "Number of transitive relations exceeded!";
+  private static final int VIRTUAL = -1;
 
-public abstract class SymbolicValueRelation {
+  static class RelationComparator implements Comparator<SymbolicValueRelation> {
 
-  protected final SymbolicValue v1;
-  protected final SymbolicValue v2;
+    @Override
+    public int compare(SymbolicValueRelation rel1, SymbolicValueRelation rel2) {
+
+      final int[] key1 = rel1.key();
+      final int[] key2 = rel2.key();
+      int answer = Integer.compare(key1[0], key2[0]);
+      if (answer == 0) {
+        answer = Integer.compare(key1[1], key2[1]);
+        if (answer == 0) {
+          answer = rel1.getClass().getName().compareTo(rel2.getClass().getName());
+        }
+      }
+      return answer;
+    }
+
+  }
+
+  public SymbolicValueRelation(int id) {
+    super(id);
+  }
 
   protected SymbolicValueRelation(SymbolicValue v1, SymbolicValue v2) {
-    this.v1 = v1;
-    this.v2 = v2;
+    super(VIRTUAL);
+    computedFrom(ImmutableList.of(v2, v1));
   }
 
   @Override
-  public int hashCode() {
-    long bits = v1.id();
-    bits ^= v2.id() * 31;
-    bits ^= getClass().hashCode() * 31;
-    return ((int) bits) ^ ((int) (bits >> 32));
+  public BooleanConstraint shouldNotInverse() {
+    return BooleanConstraint.FALSE;
   }
 
   @Override
-  public boolean equals(Object obj) {
-    if (obj instanceof SymbolicValueRelation) {
-      return equalsRelation((SymbolicValueRelation) obj);
-    }
-    return super.equals(obj);
+  public boolean isVirtual() {
+    return id() == VIRTUAL;
   }
 
-  private boolean equalsRelation(SymbolicValueRelation relation) {
-    final int[] key1 = key();
-    final int[] key2 = relation.key();
-    if (key1[0] == key2[0] && key1[1] == key2[1]) {
-      return getClass().equals(relation.getClass());
+  @Override
+  protected List<ProgramState> copyConstraint(SymbolicValue from, SymbolicValue to, ProgramState programState, BooleanConstraint booleanConstraint) {
+    ProgramState newState = programState;
+    if (programState.canReach(from) || programState.canReach(to)) {
+      newState = programState.addConstraint(this, booleanConstraint);
     }
-    return false;
+    return super.copyConstraint(from, to, newState, booleanConstraint);
   }
 
   private int[] key() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Move this method into "RelationComparator". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS3398)